File: openfeature-provider/java/src/main/java/com/spotify/confidence/sdk/UnsupportedMaterializationStore.java
package com.spotify.confidence.sdk;

import java.util.List;
import java.util.Set;
import java.util.concurrent.CompletionStage;

/**
 * A materialization store that doesn't support materializations.
 *
 * <p>This is the default materialization store. When a flag requires materialization data (sticky
 * assignments or custom targeting), the evaluation will fail with {@link
 * MaterializationNotSupportedException}.
 *
 * <p>To enable materialization support, use either:
 *
 * <ul>
 *   <li>{@link RemoteMaterializationStore} - stores data remotely via gRPC to Confidence
 *   <li>Custom implementation of {@link MaterializationStore} - stores data in your own
 *       infrastructure
 * </ul>
 */
final class UnsupportedMaterializationStore implements MaterializationStore {

  @Override
  public CompletionStage<List<ReadResult>> read(List<? extends ReadOp> ops) {
    throw new MaterializationNotSupportedException();
  }

  @Override
  public CompletionStage<Void> write(Set<? extends WriteOp> ops) {
    throw new MaterializationNotSupportedException();
  }
}

File: openfeature-provider/android/confidence-provider/src/main/kotlin/com/spotify/confidence/android/UnsupportedMaterializationStore.kt
package com.spotify.confidence.android

import java.util.concurrent.CompletableFuture

/**
 * Default materialization store that throws [MaterializationNotSupportedException]
 * for all operations, triggering fallback to remote gRPC resolution.
 */
internal class UnsupportedMaterializationStore : MaterializationStore {

    override fun read(ops: List<MaterializationStore.ReadOp>): CompletableFuture<List<MaterializationStore.ReadResult>> {
        throw MaterializationNotSupportedException()
    }

    override fun write(ops: Set<MaterializationStore.WriteOp>): CompletableFuture<Void?> {
        throw MaterializationNotSupportedException()
    }
}
